src/com/cloudogu/ces/cesbuildlib/Git.groovy

```diff
@@ -260,4 +325,88 @@ class Git implements Serializable {
             script.sh "git ${args}"
         }
     }
-}
+
+    /**
+     * Finishes a git flow release and pushes all merged branches to remote
+     *
+     * Only execute this function if you are already on a release branch
+     *
+     * @param releaseVersion the version that is going to be released
+     */
+    void finishGitRelease(String releaseVersion) {
+        String branchName = getBranchName()
+
+        // Check if tag already exists
+        if (tagExists("${releaseVersion}")){
+            error("You cannot build this version, because it already exists.")
+        }
+
+        // Make sure all branches are fetched
+        script.sh "git config 'remote.origin.fetch' '+refs/heads/*:refs/remotes/origin/*'"
+        executeGitWithCredentials("fetch --all")
+
+        // Make sure there are no changes on develop
+        if (developHasChanged(branchName)){
+            error("There are changes on develop branch that are not merged into release. Please merge and restart process.")
+        }
+
+        // Make sure any branch we need exists locally
+        script.sh "git checkout ${branchName}"
+        executeGitWithCredentials("pull origin ${branchName}")
+        script.sh "git checkout develop"
+        executeGitWithCredentials("pull origin develop")
+        script.sh "git checkout master"
+        executeGitWithCredentials("pull origin master")
+
+        // Merge release branch into master
+        script.sh "git merge --no-ff ${branchName}"
+
+        // Create tag. Use -f because the created tag will persist when build has failed.
+        executeGitWithCredentials("tag -f -m 'release version ${releaseVersion}' ${releaseVersion}")
+
+        // Merge release branch into develop
+        script.sh "git checkout develop"
+        script.sh "git merge --no-ff ${branchName}"
+
+        // Delete release branch
+        script.sh "git branch -d ${branchName}"
+
+        // Checkout tag
+        script.sh "git checkout ${releaseVersion}"
+
+        // Push changes and tags
+        executeGitWithCredentials("push origin master")
+        executeGitWithCredentials("push origin develop")
+        executeGitWithCredentials("push origin --tags")
+        executeGitWithCredentials("push origin --delete ${branchName}")
+    }
+
+    /**
+     * Creates a new release on Github and adds changelog info to it
+     *
+     * @param releaseVersion the version for the github release
+     * @param changelog the changelog object to extract the release information from
+     */
+    void createGithubRelease(String releaseVersion, Changelog changelog) {
```

Comment:
This seems far too complex to me. What will be the flow when `getChangelog()` throw an exception?
Exception is catched. Then another exception is thrown on `changelogText == ""` which is catched  some lines later.

Couldn't this be one block?

Reply:
Is `getChangelog()` supposed to throw exceptions at all?

Reply:
I reworked the method to only have one try block.
Originally we had this code block splitted into 2 SonarQube stages to make it visible when reading the changelog fails but still add  a GitHub release (with empty message). That's why we had two catch blocks.
I think the `getChangelog()` method should not throw an exception. At least i don't know any case where this would happen.